src/pages/academy/notiPreference/NotiPreference.tsx

```diff
@@ -0,0 +1,184 @@
+    }
+
+    // Initialise notification preferences if absent
+    configurableNotificationConfigs.current?.forEach(config => {
+      if (config.notificationPreference === null) {
+        config.notificationPreference = {
+          id: -1,
+          isEnabled: true,
+          timeOptionId: null
+        };
+      }
+    });
+  }, [session]);
+
+  const setIsEnabled = (index: number, value: boolean) => {
+    const temp = [...configurableNotificationConfigs.current];
+
+    temp[index]['notificationPreference'].isEnabled = value;
+
+    configurableNotificationConfigs.current = temp;
+    gridApi.current
+      ?.getDisplayedRowAtIndex(index)
+      ?.setDataValue('notificationPreference.isEnabled', value);
+    setHasChanges(true);
+  };
+
+  const setTimeOption = (index: number, value: TimeOption) => {
+    const temp = [...configurableNotificationConfigs.current];
+
+    temp[index]['notificationPreference'].timeOptionId = value.id;
+
+    configurableNotificationConfigs.current = temp;
+    gridApi.current
+      ?.getDisplayedRowAtIndex(index)
+      ?.setDataValue('timeOptions', temp[index]['timeOptions']);
+    setHasChanges(true);
+  };
+
+  const assessmentTypeFormatter: ValueFormatterFunc<NotificationConfiguration> = params => {
+    return params.data!.assessmentConfig?.type || '-';
+  };
+
+  const recipientFormatter: ValueFormatterFunc<NotificationConfiguration> = params => {
+    return params.data!.notificationType.forStaff ? 'Staff' : 'Student';
+  };
+
+  const columnDefs = [
+    {
+      headerName: 'Notification Type',
+      field: 'notificationType.name',
+      rowDrag: true
+    },
+    {
+      headerName: 'Assessment Type',
+      field: 'assessmentConfig.type',
+      valueFormatter: assessmentTypeFormatter
+    },
+    {
+      headerName: 'Recipients',
+      field: 'notificationType.forStaff',
+      valueFormatter: recipientFormatter
+    },
+    {
+      headerName: 'Reminder',
+      field: 'timeOptions',
+      cellRendererFramework: SelectCell,
+      cellRendererParams: {
+        setStateHandler: setTimeOption,
+        field: 'timeOptions'
+      }
+    },
+    {
+      headerName: 'Enabled',
+      field: 'notificationPreference.isEnabled',
+      cellRendererFramework: BooleanCell,
+      cellRendererParams: {
+        setStateHandler: setIsEnabled,
+        field: 'notificationPreference.isEnabled'
+      }
+    }
+  ];
+
+  const defaultColumnDefs = {
+    filter: false,
+    resizable: true,
+    sortable: false
+  };
+
+  const onGridReady = (params: GridReadyEvent) => {
+    gridApi.current = params.api;
+    params.api.sizeColumnsToFit();
+  };
+
+  const submitHandler = () => {
+    if (!hasChanges) return;
+
+    const preferences: NotificationPreference[] = configurableNotificationConfigs.current.map(
+      config => {
+        return {
+          ...config.notificationPreference,
+          notificationConfigId: config.id
+        } as NotificationPreference;
```

Comment:
Ditto on this typecast resulting in potential type safety issues; is there a need for this typecast in the first place?